File: src/main/java/com/turleylabs/algo/trader/kata/framework/Slice.java
package com.turleylabs.algo.trader.kata.framework;

import com.turleylabs.algo.trader.kata.framework.Bar;
import com.turleylabs.algo.trader.kata.framework.CBOE;

import java.io.*;
import java.time.LocalDate;

public class Slice {
    private final LocalDate date;

    public Slice(LocalDate tradeDate) {
        this.date = tradeDate;
    }

    private Bar findBar() {
        ClassLoader classLoader = getClass().getClassLoader();
        try (BufferedReader br = new BufferedReader(new FileReader(classLoader.getResource("TQQQ.csv").getFile()))) {
            String line = br.readLine();
            while ((line = br.readLine()) != null) {
                String[] values = line.split(",");
                if (LocalDate.parse(values[0]).equals(this.date)) {
                    return new Bar(Double.parseDouble(values[5]));
                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public boolean ContainsKey(String symbol) {
        return findBar() == null;
    }

    public CBOE getCBOE(String symbol) {
        return findCBOE();
    }

    private CBOE findCBOE() {
        ClassLoader classLoader = getClass().getClassLoader();
        try (BufferedReader br = new BufferedReader(new FileReader(classLoader.getResource("VIX.csv").getFile()))) {
            String line = br.readLine();
            while ((line = br.readLine()) != null) {
                String[] values = line.split(",");
                if (LocalDate.parse(values[0]).equals(this.date)) {
                    return new CBOE(Double.parseDouble(values[5]));
                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public Bar get(String symbol) {
        return findBar();
    }

    public LocalDate getDate() {
        return date;
    }
}

File: typescript/src/framework/slice.ts
import { readFileSync } from 'fs';
import { join } from 'path';
import { CBOE } from './cboe';
import { Bar } from './bar';

export class Slice {
  constructor(readonly date: Date, private symbol: string) {
  }

  containsKey(symbol: String): boolean {
    return symbol === this.symbol;
  }

  getCBOE(symbol: string): CBOE {
    if (!this.containsKey(symbol)) {
      return null;
    }
    return this.findCBOE();
  }

  get(symbol: string): Bar {
    if (!this.containsKey(symbol)) {
      return null;
    }
    return this.findBar(symbol);
  }

  private findCBOE(): CBOE {
    try {
      const fileContents = readFileSync(join(__dirname, '..', '..', 'data', 'VIX.csv'), { encoding: 'utf8'});
      const lines = fileContents.split('\n');
      lines.splice(0, 1);
      for (const line of lines) {
        const values = line.split(',');
        const date = new Date(Date.parse(values[0]));
        if (
          date.toISOString().split('T')[0] ===
          this.date.toISOString().split('T')[0]
        ) {
          return new CBOE(parseFloat(values[5]));
        }
      }
    } catch (e) {
      console.error(e);
    }
  
    return null;
  }

  private findBar(symbol: string): Bar {
    try {
      const fileContents = readFileSync(join(__dirname, '..', '..', 'data', `${symbol}.csv`), { encoding: 'utf8'});
      const lines = fileContents.split('\n');
      lines.splice(0, 1);
      for (const line of lines) {
        const values = line.split(',');
        const date = new Date(Date.parse(values[0]));
        if (
          date.toISOString().split('T')[0] === 
          this.date.toISOString().split('T')[0]
        ) {
          return new Bar(parseFloat(values[5]));
        }
      }
    } catch (e) {
      console.error(e);
    }
    return null;
  }
}